### PR TITLE
Wrong user flow involving analyses and empty layers

### LINKED ISSUE #11876: Improve onboarding for when user adds an empty layer
### Context
Design ticket https://github.com/CartoDB/design/issues/612

### Additional info
Here's the Invision Project (linked)
https://invis.io/6QAZN3JSR

### PR DESCRIPTION
**Pull request by jesusbotella**
This PR fixes `Wrong user flow involving analyses and empty layers` issue. The code in this PR enables the analysis tab when layer is empty and has an analysis.

I have also re-implemented the feature that we had to rollback in https://github.com/CartoDB/cartodb/pull/13432 PR, showing the warning icon when a layer is empty in the DataLayer view. This finishes the empty layer improvement, so that we can close #11876.

Related to #13478 and #13486.

Acceptance
-----
* User: 13489-wrong-user-flow

#### Analysis tab enabled when it has filtered out all the data

1. Apply an analysis in a layer that will filter out all the data in that layer
2. Go back to the Layers list
3. Click on the previous layer again
4. Check that Analysis tab is enabled

#### Empty layer icon in the layer list
1. Create an empty layer
2. Check that the warning icon is shown in the Layers list
<br/>

3. Click on the layer
4. Add any geometry to the layer
5. Go back to the Layers list
6. Check that the warning icon is gone
<br/>

7. Apply an analysis in a layer that will filter out all the data in that layer
8. Go back to the Layers list
9. Check that the warning icon is shown in the Layers list
<br/>

10. Remove the previously added analysis,
11. Apply another analysis in the layer without filtering out all the data in the layer
12. Go back to the Layers list
13. Check that the warning icon is gone

**Test any other use case you can think of that could be affected by these changes 😀**

### CODE REVIEW

**Comment by jesusbotella** [comment]
retest this please

**Comment by matallo** [comment]
I know `layer-content-model.spec.js` doesn't exist yet (my fault because I created the model without tests), but at least let's add a spec with the new methods

File: lib/assets/javascripts/cartodb3/data/layer-definition-model.js [hunk 1]
```diff
@@ -297,6 +297,10 @@ module.exports = Backbone.Model.extend({
       layerTypesAndKinds.isTorqueType(layerType);
   },
 
+  isVisible: function () {
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/data/layer-definition-model.js]: not covered in the tests

File: lib/assets/javascripts/cartodb3/data/query-geometry-model.js [hunk 2]
```diff
@@ -99,19 +99,23 @@ module.exports = BaseModel.extend({
     return Backbone.Model.prototype.fetch.call(this, opts);
   },
 
-  parse: function (r) {
-    var simpleGeom;
-
-    _.some(r.rows, function (row) {
-      return !!(simpleGeom = row.the_geom); // to stop when found a valid simple geom
-    });
+  parse: function (data) {
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/data/query-geometry-model.js]: not covered in the tests

File: lib/assets/javascripts/cartodb3/data/query-geometry-model.js [hunk 3]
```diff
@@ -99,19 +99,23 @@ module.exports = BaseModel.extend({
...
     return {
       status: STATUS.fetched,
-      simple_geom: simpleGeom || ''
+      simple_geom: simpleGeom && simpleGeom.the_geom
+        ? simpleGeom.the_geom
+        : ''
     };
   },
 
+  _findValidSimpleGeom: function (data) {
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/data/query-geometry-model.js]: not covered in the tests

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-view.js [hunk 4]
```diff
@@ -468,24 +483,32 @@ module.exports = CoreView.extend({
     return this.model.get('context') === CONTEXTS.TABLE;
   },
 
+  _isLayerEmpty: function () {
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-view.js]: not covered in the tests

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-view.js [hunk 5]
```diff
@@ -468,24 +483,32 @@ module.exports = CoreView.extend({
...
+    var analysisDefinitionNodeModel = this._getAnalysisDefinitionNodeModel();
+
+    return (analysisDefinitionNodeModel.isReadOnly() ||
+      !this._layerDefinitionModel.get('visible'));
+  },
+
+  _isDataEmpty: function () {
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-view.js]: not covered in the tests

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-content-view.js [hunk 6]
```diff
@@ -105,10 +108,23 @@ module.exports = CoreView.extend({
     );
   },
 
+  _showEmptyData: function () {
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-content-view.js]: not covered in the tests

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-content-view.js [hunk 7]
```diff
@@ -105,10 +108,23 @@ module.exports = CoreView.extend({
...
+        action: _t('editor.messages.no-data.action-message')
+      })
+    );
+  },
+
   _isErrored: function () {
     return this._layerContentModel.isErrored();
   },
 
+  _isNoGeometryData: function () {
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-content-view.js]: not covered in the tests

File: lib/assets/javascripts/cartodb3/data/layer-definition-model.js (hunk 1 above)
Comment (elenatorro) [reply lib/assets/javascripts/cartodb3/data/layer-definition-model.js]: 👍

Comment (matallo) [reply lib/assets/javascripts/cartodb3/data/layer-definition-model.js]: still pending!

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-view.js (hunk 5 above)
Comment (matallo) [reply lib/assets/javascripts/cartodb3/editor/layers/layer-content-view.js]: pending!

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-content-view.js (hunk 6 above)
Comment (matallo) [reply lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-content-view.js]: 👍

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-content-view.js (hunk 7 above)
Comment (matallo) [reply lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-content-view.js]: pending!

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js [hunk 8]
```diff
@@ -425,6 +426,19 @@ module.exports = DatasetBaseView.extend({
     setTimeout(this._redirectIfErrors.bind(this), 0);
   },
 
+  _createControlView: function () {
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js]: pending!

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js [hunk 9]
```diff
@@ -425,6 +426,19 @@ module.exports = DatasetBaseView.extend({
...
+    return new Toggler({
+      model: this._togglerModel
+    });
+  },
+
+  _createActionView: function () {
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js]: pending!

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindows-view.js [hunk 10]
```diff
@@ -54,7 +58,16 @@ module.exports = CoreView.extend({
     this.listenTo(this._layerTooltipModel, 'change:template', this._updateSelectedChild);
   },
 
-  _initViews: function () {
+  _showEmptyData: function () {
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindows-view.js]: pending!

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindows-view.js [hunk 11]
```diff
@@ -54,7 +58,16 @@ module.exports = CoreView.extend({
...
+    this.$el.append(
+      layerGenericMessageTemplate({
+        message: _t('editor.messages.no-data.message'),
+        action: _t('editor.messages.no-data.action-message')
+      })
+    );
+  },
+
+  _renderInfowindowMenu: function () {
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindows-view.js]: pending!

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/legend/legends-view.js [hunk 12]
```diff
@@ -119,6 +118,15 @@ module.exports = CoreView.extend({
     this._changeStyle(this._editorModel);
   },
 
+  _showEmptyData: function () {
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/legend/legends-view.js]: pending!

File: lib/assets/javascripts/cartodb3/editor/style/style-content-view.js [hunk 13]
```diff
@@ -123,6 +127,15 @@ module.exports = CoreView.extend({
     this._editorModel.set({ disabled: this._isErrored() || !hasGeometry });
   },
 
+  _showEmptyData: function () {
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/style/style-content-view.js]: pending!

File: lib/assets/javascripts/cartodb3/editor/style/style-view.js [hunk 14]
```diff
@@ -211,23 +220,27 @@ module.exports = CoreView.extend({
...
-      }
+      createControlView: createControlView,
+      createActionView: createActionView
     });
 
     this.$el.append(panelWithOptionsView.render().el);
     this.addView(panelWithOptionsView);
   },
 
+  _createControlView: function () {
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/style/style-view.js]: pending!

File: lib/assets/javascripts/cartodb3/editor/style/style-view.js [hunk 15]
```diff
@@ -211,23 +220,27 @@ module.exports = CoreView.extend({
...
+    return new Toggler({
+      model: this._togglerModel
+    });
+  },
+
+  _createActionView: function () {
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/style/style-view.js]: pending!

File: lib/assets/javascripts/cartodb3/data/query-geometry-model.js (hunk 2 above)
Comment (matallo) [reply lib/assets/javascripts/cartodb3/data/query-geometry-model.js]: 👍

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-view.js (hunk 4 above)
Comment (matallo) [reply lib/assets/javascripts/cartodb3/editor/layers/layer-content-view.js]: 👍

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/legend/legends-view.js [hunk 16]
```diff
@@ -45,6 +46,8 @@ module.exports = CoreView.extend({
 
     if (this._isErrored()) {
       this._renderError();
+    } else if (this._isDataEmpty) {
+      this._showEmptyData();
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/legend/legends-view.js]: this is not covered!

File: lib/assets/javascripts/cartodb3/data/layer-definition-model.js (hunk 1 above)
Comment (matallo) [reply lib/assets/javascripts/cartodb3/data/layer-definition-model.js]: 👍

File: lib/assets/javascripts/cartodb3/data/query-geometry-model.js (hunk 3 above)
Comment (matallo) [reply lib/assets/javascripts/cartodb3/data/query-geometry-model.js]: I'm blind 🤦‍♂️

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-view.js (hunk 5 above)
Comment (matallo) [reply lib/assets/javascripts/cartodb3/editor/layers/layer-content-view.js]: 👍

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-content-view.js (hunk 7 above)
Comment (elenatorro) [reply lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-content-view.js]: done 👌

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js (hunk 8 above)
Comment (elenatorro) [reply lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js]: done 👌

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindows-view.js (hunk 11 above)
Comment (elenatorro) [reply lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/infowindow/infowindows-view.js]: done 👌

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/legend/legends-view.js (hunk 16 above)
Comment (elenatorro) [reply lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/legend/legends-view.js]: done 👌

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/legend/legends-view.js (hunk 12 above)
Comment (elenatorro) [reply lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/legend/legends-view.js]: done 👌

File: lib/assets/javascripts/cartodb3/editor/style/style-content-view.js (hunk 13 above)
Comment (elenatorro) [reply lib/assets/javascripts/cartodb3/editor/style/style-content-view.js]: done 👌

File: lib/assets/javascripts/cartodb3/editor/style/style-view.js (hunk 14 above)
Comment (elenatorro) [reply lib/assets/javascripts/cartodb3/editor/style/style-view.js]: done 👌

**Comment by matallo** [comment]
retest this, please

**Comment by ivanmalagon** [comment]
There are contradictory acceptance steps between the PR description and [this comment](https://github.com/CartoDB/cartodb/issues/13478#issuecomment-363125888) of @arianaescobar .
She says that a layer that has the data filtered out is not empty so no icon should be displayed. In the acceptance tests, it says that the icon must be displayed.

In the current solution, the icon is shown as you can see below, contrary of what's specified by Ari. What is the desired behaviour?

image

**Comment by elenatorro** [comment]
You are right, @ivanmalagon. If there is no data, the "No data available" message should be shown, but not the icon.

**Comment by arianaescobar** [comment]
Yes, please remove the icon 🙏🏼🙃

I owe you a hug, @ivanmalagon

**Comment by arianaescobar** [comment]
LGTM! ✨

**Pull request closed without merging**